### ISSUE TITLE
finding the wrong records / rails 4.2

### ISSUE THREAD

**Issue by mmank** [composite-primary-keys/composite_primary_keys#291]
TLDR: this gem finds the wrong record

I have a simple class structure:

    class Department < ActiveRecord::Base
	self.primary_key = :aid, :bid
    end
    class Employee < ActiveRecord::Base
	belongs_to :department, foreign_key: [:aid, :bid]
    end

the tables:

		    create_table :employees do |t|
		      t.integer :aid, null:false
		      t.integer :bid, null:false
		      t.string :name

		      t.timestamps null: false
		    end

		    create_table :departments, id: false do |t|
		      t.integer :aid, null:false
		      t.integer :bid, null:false
		      t.string :deptname

		      t.timestamps null: false
		    end

		    execute 'alter table departments add primary key(aid, bid)'

Link to sample app: https://www.dropbox.com/s/i90ja8pqbogmr0l/foo.zip?dl=0 . This is Rails 4.2.0 .

Look at the log below. Notice how CPK finds the wrong department. It somehow caches the first department found.

        irb(main):001:0> Department.create!(aid:1, bid:2, deptname: 'd1')
           (0.1ms)  BEGIN
          SQL (0.5ms)  INSERT INTO "departments" ("aid", "bid", "deptname", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "aid","bid"  [["aid", 1], ["bid", 2], ["deptname", "d1"], ["created_at", "2015-03-07 11:44:56.119700"], ["updated_at", "2015-03-07 11:44:56.119700"]]
           (0.3ms)  COMMIT
        => #<Department aid: 1, bid: 2, deptname: "d1", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):002:0> Department.create!(aid:3, bid:4, deptname: 'd2')
           (0.1ms)  BEGIN
          SQL (0.2ms)  INSERT INTO "departments" ("aid", "bid", "deptname", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "aid","bid"  [["aid", 3], ["bid", 4], ["deptname", "d2"], ["created_at", "2015-03-07 11:44:56.126599"], ["updated_at", "2015-03-07 11:44:56.126599"]]
           (0.3ms)  COMMIT
        => #<Department aid: 3, bid: 4, deptname: "d2", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):003:0> Employee.create!(aid:1, bid: 2, name:'e1')
           (0.1ms)  BEGIN
          SQL (6.1ms)  INSERT INTO "employees" ("aid", "bid", "name", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["aid", 1], ["bid", 2], ["name", "e1"], ["created_at", "2015-03-07 11:44:56.144813"], ["updated_at", "2015-03-07 11:44:56.144813"]]
           (4.8ms)  COMMIT
        => #<Employee id: 1, aid: 1, bid: 2, name: "e1", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):004:0> Employee.create!(aid:3, bid: 4, name:'e2')
           (0.1ms)  BEGIN
          SQL (0.2ms)  INSERT INTO "employees" ("aid", "bid", "name", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["aid", 3], ["bid", 4], ["name", "e2"], ["created_at", "2015-03-07 11:44:56.159133"], ["updated_at", "2015-03-07 11:44:56.159133"]]
           (0.2ms)  COMMIT
        => #<Employee id: 2, aid: 3, bid: 4, name: "e2", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):005:0> Employee.find(1).department
          Employee Load (0.2ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = $1 LIMIT 1  [["id", 1]]
          Department Load (0.4ms)  SELECT  "departments".* FROM "departments" WHERE ("departments"."aid" = 1 AND "departments"."bid" = 2) LIMIT 1
        => #<Department aid: 1, bid: 2, deptname: "d1", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):006:0> Employee.find(2).department
          Employee Load (0.3ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = $1 LIMIT 1  [["id", 2]]
          Department Load (0.3ms)  SELECT  "departments".* FROM "departments" WHERE ("departments"."aid" = 1 AND "departments"."bid" = 2) LIMIT 1
        => #<Department aid: 1, bid: 2, deptname: "d1", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):007:0> reload!
        Reloading...
        => true
        irb(main):008:0> Employee.find(2).department
          Employee Load (0.3ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = $1 LIMIT 1  [["id", 2]]
          Department Load (0.2ms)  SELECT  "departments".* FROM "departments" WHERE ("departments"."aid" = 3 AND "departments"."bid" = 4) LIMIT 1
        => #<Department aid: 3, bid: 4, deptname: "d2", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):009:0> Employee.find(1).department
          Employee Load (0.5ms)  SELECT  "employees".* FROM "employees" WHERE "employees"."id" = $1 LIMIT 1  [["id", 1]]
          Department Load (0.4ms)  SELECT  "departments".* FROM "departments" WHERE ("departments"."aid" = 3 AND "departments"."bid" = 4) LIMIT 1
        => #<Department aid: 3, bid: 4, deptname: "d2", created_at: "2015-03-07 11:44:56", updated_at: "2015-03-07 11:44:56">
        irb(main):010:0>

**Comment by mmank** [comment]
Hmm, this seems to be fixed in ar_4.2.x branch tip. I think this warrants a new release...

**Comment by cfis** [comment]
Sorry for the delay, will cut a new release.
